code/autoupdater/autoupdater.c

```diff
@@ -0,0 +1,780 @@
+    int64_t len;
+    int update;
+    int rollback;
+    struct ManifestItem *next;
+} ManifestItem;
+
+static ManifestItem *manifest = NULL;
+
+static void freeManifest(void)
+{
+    ManifestItem *item = manifest;
+    manifest = NULL;
+
+    while (item != NULL) {
+        ManifestItem *next = item->next;
+        free(item->fname);
+        free(item);
+        item = next;
+    }
+    manifest = NULL;
+}
+
+static const char *timestamp(void)
+{
+    time_t t = time(NULL);
+    char *retval = asctime(localtime(&t));
+    if (retval) {
+        char *ptr;
+        for (ptr = retval; *ptr; ptr++) {
+            if ((*ptr == '\r') || (*ptr == '\n')) {
+                *ptr = '\0';
+                break;
+            }
+        }
+    }
+    return retval ? retval : "[date unknown]";
+}
+
+
+static FILE *logfile = NULL;
+
+#define SDL_PRINTF_VARARG_FUNC( fmtargnumber ) __attribute__ (( format( __printf__, fmtargnumber, fmtargnumber+1 )))
+
+static void info(const char *str)
+{
+    fputs(str, logfile);
+    fputs("\n", logfile);
+    fflush(logfile);
+}
+
+static void infof(const char *fmt, ...)
+#if defined(__GNUC__) || defined(__clang__)
+__attribute__ (( format( __printf__, 1, 2 )))
+#endif
+;
+
+static void infof(const char *fmt, ...)
+{
+    va_list ap;
+    va_start(ap, fmt);
+    vfprintf(logfile, fmt, ap);
+    va_end(ap);
+    fputs("\n", logfile);
+    fflush(logfile);
+}
+
+static void restoreRollbacks(void)
+{
+    /* you can't call die() in this function! If this fails, you're doomed. */
+    ManifestItem *item;
+    for (item = manifest; item != NULL; item = item->next) {
+        if (item->rollback) {
+            char rollbackPath[64];
+            snprintf(rollbackPath, sizeof (rollbackPath), "updates/rollbacks/%d", item->rollback);
+            infof("restore rollback: '%s' -> '%s'", rollbackPath, item->fname);
+            remove(item->fname);
+            rename(rollbackPath, item->fname);
+        }
+    }
+}
+
+static void die(const char *why) NEVER_RETURNS;
+
+
+#ifndef _WIN32  /* hooray for Unix linker hostility! */
+#undef curl_easy_setopt
+#include <dlfcn.h>
+typedef void (*CURLFN_curl_easy_cleanup)(CURL *curl);
+typedef CURL *(*CURLFN_curl_easy_init)(void);
+typedef CURLcode (*CURLFN_curl_easy_setopt)(CURL *curl, CURLoption option, ...);
+typedef CURLcode (*CURLFN_curl_easy_perform)(CURL *curl);
+typedef CURLcode (*CURLFN_curl_global_init)(long flags);
+typedef void (*CURLFN_curl_global_cleanup)(void);
+
+static CURLFN_curl_easy_cleanup CURL_curl_easy_cleanup;
+static CURLFN_curl_easy_init CURL_curl_easy_init;
+static CURLFN_curl_easy_setopt CURL_curl_easy_setopt;
+static CURLFN_curl_easy_perform CURL_curl_easy_perform;
+static CURLFN_curl_global_init CURL_curl_global_init;
+static CURLFN_curl_global_cleanup CURL_curl_global_cleanup;
+
+static void load_libcurl(void)
+{
+    #ifdef __APPLE__
+    const char *libname = "libcurl.4.dylib";
+    #else
+    const char *libname = "libcurl.so.4";
```

Comment:
This will fail on OpenBSD

/usr/local/lib/libcurl.so.25.9 (value from OpenBSD 6.1)